Sources/Core/KeystoreManager/BIP39.swift

```diff
@@ -69,50 +69,88 @@ public enum BIP39Language {
-            wordList.append(word)
+    static public func generateMnemonics(entropy: Int, language: BIP39Language = .english) -> [String]? {
+        guard let entropy = entropyOf(size: entropy) else { return nil }
+        return generateMnemonicsFrom(entropy: entropy, language: language)
+    }
+
+    static private func entropyOf(size: Int) -> Data? {
+        guard size >= 128 && size <= 256 && size.isMultiple(of: 32) else {
+            return nil
         }
+
+        return Data.randomBytes(length: size/8)
+    }
+
+    static func bitarray(from data: Data) -> String {
+        data.map {
+            let binary = String($0, radix: 2)
+            let padding = String(repeating: "0", count: 8 - binary.count)
+            return padding + binary
+        }.joined()
+    }
+
+    static func generateChecksum(entropyBytes inputData: Data, checksumLength: Int) -> String? {
+        guard let checksumData = inputData.sha256().bitsInRange(0, checksumLength) else {
+            return nil
+        }
+        let checksum = String(checksumData, radix: 2).leftPadding(toLength: checksumLength, withPad: "0")
+        return checksum
+    }
+
+    static public func generateMnemonicsFromEntropy(entropy: Data, language: BIP39Language = .english) -> String? {
+        guard entropy.count >= 16, entropy.count & 4 == 0 else {return nil}
         let separator = language.separator
+        let wordList = generateMnemonicsFrom(entropy: entropy)
         return wordList.joined(separator: separator)
     }
 
-    /// Initializes a new mnemonics set with the provided bitsOfEntropy.
-    /// - Parameters:
-    ///   - bitsOfEntropy: 128 - 12 words, 192 - 18 words , 256 - 24 words in output.
-    ///   - language: words language, default english
-    /// - Returns: random 12-24 words, that represent new Mnemonic phrase.
-    static public func generateMnemonics(bitsOfEntropy: Int, language: BIP39Language = BIP39Language.english) throws -> String? {
-        guard bitsOfEntropy >= 128 && bitsOfEntropy <= 256 && bitsOfEntropy.isMultiple(of: 32) else {return nil}
-        guard let entropy = Data.randomBytes(length: bitsOfEntropy/8) else {throw AbstractKeystoreError.noEntropyError}
-        return BIP39.generateMnemonicsFromEntropy(entropy: entropy, language:
-        language)
+    static public func generateMnemonicsFrom(entropy: Data, language: BIP39Language = .english) -> [String]  {
+        let entropyBitSize = entropy.count * 8
+        let checksum_length = entropyBitSize / 32
+
+        var entropy_bits = bitarray(from: entropy)
 
+        guard let checksumTest = generateChecksum(entropyBytes: entropy, checksumLength: checksum_length) else {
+            return []
+        }
+        entropy_bits += checksumTest
+        return entropy_bits
+            .split(every: 11)
+            .compactMap { binary in
+            Int(binary, radix: 2)
+        }
+        .map { index in
+            language.words[index]
+        }
```

Comment:
Formatting is required.
Please make sure all files you worked with are formatted.